src/Dijkstra/AST/Examples/Maybe/Branching.agda

```diff
@@ -16,33 +16,196 @@ module Dijkstra.AST.Examples.Maybe.Branching where
+
+    C-c C-, shows that the second hole needs evidence that n + (n + zero) ≡ n + (n + zero), which is
+    easily satisfied by refl, resulting in the complete proof below.
+
+    One important lesson is to remember to use the variants of C-c C-, with zero, one or two C-u's
+    before it.  Each is the most useful of the three in some contexts.  Experienced users might
+    predict which one is best, but just having enough experience to remember to try all three is a
+    big help.
+
+    -}
+
+    proj₁ (branchingProgWP _) isTrue  =           toWitnessT isTrue
+    proj₂ (branchingProgWP _) isFalse = (n≢0⇒n>0 (toWitnessF isFalse)) , refl
+
+    -- And therefore, the result of running the program satisfies the postcondition
+    prop : (i : Input) → bpPost (runMaybeExt branchingProg i)
+    prop i =
+      ASTSufficientPT.sufficient MaybeSufExt branchingProg bpPost i (branchingProgWP i)
+
+  module Prettier where
+    open BranchingSyntax MaybeOps
+    open MaybeBranchingSyntax
+
+    -- Same program with nicer syntax using ifAST, bail and return
+    branchingProg : MaybeDExt ℕ
+    branchingProg = ifAST ⌊ n ≟ℕ 0 ⌋
+                    then bail
+                    else (return (2 * n))
+
+    --  Note that the same proof works for both versions (as they are equivalent)
+    branchingProgWP : (i : Input)
+                         → ASTPredTrans.predTrans MaybePTExt branchingProg bpPost i
+    proj₁ (branchingProgWP i) isTrue  =           toWitnessT isTrue
+    proj₂ (branchingProgWP i) isFalse = (n≢0⇒n>0 (toWitnessF isFalse)) , refl
+
+    prop : (i : Input) → bpPost (runMaybeExt branchingProg i)
+    prop i = ASTSufficientPT.sufficient MaybeSufExt branchingProg bpPost i (branchingProgWP i)
+
+module Example-either (n : ℕ) where
+  open BranchingSyntax MaybeOps
+  open MaybeBranchingSyntax
+
+  module Common where
+
+    _monus1 : ℕ → Either Unit ℕ
+    _monus1 0        = Left unit
+    _monus1 (suc n') = Right n'
+
+    monus1lemma1 : ∀ {n'} → n ≡ n' → ∀ {l} → n' monus1 ≡ Left l → n ≡ 0
+    monus1lemma1 {n'} n≡n' _
+       with n'
+    ...| 0 = n≡n'
+
+    monus1lemma2 : ∀ {nalias} → n ≡ nalias → ∀ {n'} → nalias monus1 ≡ Right n' → n ≡ suc n'
+    monus1lemma2 {nalias} n≡nalias isrgt
+       with nalias
+    ...| suc x rewrite n≡nalias | inj₂-injective isrgt = refl
+
+    bpPost : Post ℕ
+    bpPost nothing  = n ≡ 0
+    bpPost (just x) = n ≡ suc x
+
+  module Raw where
+    open Common
+    -- A branching program that bails if n monus1 is Left _
+    -- and returns b if n monus1 is Right b
+    branchingProg : MaybeDExt ℕ
+    branchingProg = ASTop (Right (BCeither (n monus1)))
+                          λ { (lift (Left  a)) → bail
+                            ; (lift (Right b)) → return b
+                            }
+
+    branchingProgWP : (i : Input)
+                      → ASTPredTrans.predTrans MaybePTExt branchingProg bpPost i
+    proj₁ (branchingProgWP i) l islft = monus1lemma1 refl islft
```

Comment:
Something I like to do is only give variable names to things that are used.
Otherwise use `_`.

Reply:
DONE